File: iosApp/iosApp/nfc/NfcAvailability.swift
import Foundation
#if canImport(CoreNFC)
import CoreNFC
#endif

/// Categorical NFC capability state for UI display.
///
/// Mirrors the Kotlin `NfcAvailabilityStatus` sealed catalogue used by
/// the Android composeApp sample. iOS only distinguishes "available"
/// vs "unavailable" because CoreNFC has no "user disabled" state — the
/// hardware either supports `NFCTagReaderSession` or it does not.
public enum NfcAvailabilityStatus: Sendable, Equatable {
    /// The device exposes `NFCTagReaderSession.readingAvailable == true`.
    /// Ready to start a contactless read.
    case available
    /// Either the simulator (no CoreNFC), an iPad without NFC, or a
    /// device whose CoreNFC entitlement check failed at install time.
    case unavailable
}

/// Abstraction over `NFCTagReaderSession.readingAvailable` so the
/// view-model can be unit-tested without a real device. Production
/// implementation is ``SystemNfcAvailability``; tests inject a fake
/// matching this protocol.
public protocol NfcAvailability: Sendable {
    /// Snapshot the current device NFC capability state.
    func current() -> NfcAvailabilityStatus
}

/// Production ``NfcAvailability`` backed by
/// `NFCTagReaderSession.readingAvailable`.
///
/// On the iOS simulator (or any target without CoreNFC) the type still
/// compiles but always reports ``NfcAvailabilityStatus/unavailable`` —
/// the sample app remains buildable on the simulator while the real
/// read flow only runs on device.
public struct SystemNfcAvailability: NfcAvailability {

    /// Build the production NFC capability checker. Trivial — kept as
    /// an explicit initializer so callers in `iOSApp` make the
    /// dependency wiring readable.
    public init() {}

    public func current() -> NfcAvailabilityStatus {
        #if canImport(CoreNFC)
        return NFCTagReaderSession.readingAvailable ? .available : .unavailable
        #else
        return .unavailable
        #endif
    }
}

File: composeApp/src/androidMain/kotlin/io/github/a7asoft/nfc/NfcAvailability.kt
package io.github.a7asoft.nfc

import android.content.Context
import android.nfc.NfcAdapter
import android.nfc.NfcManager

/**
 * Categorical NFC capability state for UI display.
 *
 * Mapped from the platform `NfcAdapter` API; abstracted via [NfcAvailability]
 * so the ViewModel can be unit-tested without Robolectric or a real device.
 */
public sealed interface NfcAvailabilityStatus {

    /** Device has no NFC hardware. App cannot read cards. */
    public data object Unavailable : NfcAvailabilityStatus

    /** NFC hardware exists but is disabled in system settings. User can fix. */
    public data object Disabled : NfcAvailabilityStatus

    /** NFC hardware exists and is enabled. Ready to read. */
    public data object Available : NfcAvailabilityStatus
}

/**
 * Abstraction over platform NFC capability checks. Production impl is
 * [SystemNfcAvailability]; tests inject a fake.
 */
public fun interface NfcAvailability {
    /** Snapshot the current device NFC capability state. */
    public fun current(): NfcAvailabilityStatus
}

/**
 * Production [NfcAvailability] backed by [NfcAdapter].
 *
 * Captures the [Context] only as an Application context so it cannot
 * leak Activity references. Re-checks on every [current] call so the
 * ViewModel can see system-settings changes after `onResume`.
 */
public class SystemNfcAvailability(context: Context) : NfcAvailability {

    private val applicationContext: Context = context.applicationContext

    override fun current(): NfcAvailabilityStatus {
        // why: NfcAdapter.getDefaultAdapter(Context) is deprecated since
        // API 33; use the NfcManager system service. minSdk 24 covers
        // getSystemService(Class).
        val adapter = applicationContext.getSystemService(NfcManager::class.java)?.defaultAdapter
            ?: return NfcAvailabilityStatus.Unavailable
        return if (adapter.isEnabled) NfcAvailabilityStatus.Available else NfcAvailabilityStatus.Disabled
    }
}
